[podmanreceiver] Add metrics and resource metadata

rogercoll opened the pull request:
**Description:** <Describe what has changed.>
- Adds "metadata.yml" file to autogenerate metrics and resources.
- [Update: not done in this PR] Fixes invalid network metrics: "rx -> input" and "tx -> output"

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/28640

**Testing:** Previous tests preserved.

**Documentation:** <Describe the documentation added.>

rogercoll committed:
feat: add Podman receiver metadata

rogercoll committed:
docs: add change log file

rogercoll committed:
fix: gci linter

rogercoll committed:
fix: tidy module file

rogercoll committed:
fix: spelling mistake

rogercoll committed:
fix: allow container duplicated metrics

fatsheep9146 commented on receiver/podmanreceiver/metadata.yaml:
```diff
@@ -8,3 +8,115 @@ status:
+  container.blockio.io_service_bytes_recursive.read:
+    enabled: true
+    description: "Number of bytes transferred from the disk by the container"
+    extended_documentation: "[More docs]i/www.kernel.org/doc/Documentation/cgroup-v1/blkio-controller.txt)."
```
It seems that the format of link is not valid. Should it be like this? [More docs](https://www.kernel.org/doc/Documentation/cgroup-v1/blkio-controller.txt).

rogercoll replied:
Fixed in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/30232/commits/df281be2f8410406c032f3c55dfa8e39a931cd44

fatsheep9146 commented:
Could you also update the `metric` section in ReadMe.md to let user to see the `documentation.md` to check about the metrics supported by podman receiver?

fatsheep9146 commented on receiver/podmanreceiver/metadata.yaml:
```diff
@@ -8,3 +8,115 @@ status:
+  container.memory.usage.limit:
+    enabled: true
+    description: "Memory limit of the container."
+    unit: 1
```
I think the memory related metrics' unit should be `By`.

fatsheep9146 replied:
You can take https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/hostmetricsreceiver/internal/scraper/memoryscraper/metadata.yaml as example

rogercoll replied:
Makes sense, fixed in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/30232/commits/df281be2f8410406c032f3c55dfa8e39a931cd44

rogercoll committed:
fix: use "By" unit type for memory metrics

rogercoll commented:
> Could you also update the `metric` section in ReadMe.md to let user to see the `documentation.md` to check about the metrics supported by podman receiver?

Definitely, thanks for the feedback. Added in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/30232/commits/df281be2f8410406c032f3c55dfa8e39a931cd44

rogercoll committed:
Merge branch 'main' into add_podman_metadata

fatsheep9146 commented on receiver/podmanreceiver/metadata.yaml:
```diff
@@ -8,3 +8,119 @@ status:
+  container.blockio.io_service_bytes_recursive.write:
+    enabled: true
+    description: "Number of bytes transferred to the disk by the container"
+    extended_documentation: "[More docs]i/www.kernel.org/doc/Documentation/cgroup-v1/blkio-controller.txt)."
```
This line should also be fixed

rogercoll replied:
Fixed in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/30232/commits/5f3528cebd4e3ec56b485b9f989643c921319754. Thanks

rogercoll committed:
docs: fix blkio metric kernel documentation url

fatsheep9146 commented on receiver/podmanreceiver/metadata.yaml:
```diff
@@ -8,3 +8,119 @@ status:
+  container.cpu.usage.system:
+    enabled: true
+    description: "System CPU usage."
+    unit: ns
```
https://opentelemetry.io/docs/specs/semconv/system/system-metrics/#metric-systemcputime
I'm not sure if cputime unit should be `ns` or `s`, @open-telemetry/collector-approvers should this metric use `s` as unit?

rogercoll replied:
I think that cputime metric's unit is in seconds, as this is how many OSes report it (/proc/stats). But containers are handled by cgroup controllers, which allow for [better precision](https://www.kernel.org/doc/Documentation/cgroup-v2.txt). I would say not to lose precision, instead add container metrics into the semantic convention.

fatsheep9146 replied:
Your comments also make sense, I will pull this discussion to slacks, to get more input. @rogercoll

mackjmr replied:
The docker stats receiver collects `container.cpu.usage.system` as well: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/dockerstatsreceiver/metadata.yaml#L63-L67, and the unit is `ns`. IMO we should also use `ns` here to stay consistent between the two.

rogercoll replied:
Sounds good! Does it make sense to start working on the container's semantic convention PR? (feel free to add me into the slack thread _@rogercoll_)

fatsheep9146 replied:
No one responses for that for a day. And as @mackjmr  said docker stats receiver also take `ns` as unit, so I think this is ok to approve.

mackjmr commented on cmd/mdatagen/validate_test.go:
```diff
@@ -108,9 +108,17 @@ func TestValidate(t *testing.T) {
-		"container.cpu.utilization": {"docker_stats", "kubeletstats"},
-		"container.memory.rss":      {"docker_stats", "kubeletstats"},
-		"container.uptime":          {"docker_stats", "kubeletstats"},
+		"container.cpu.utilization":           {"docker_stats", "kubeletstats", "podman_stats"},
```
```suggestion
		"container.cpu.utilization":           {"docker_stats", "kubeletstats"},
```

rogercoll replied:
Thanks!

rogercoll committed:
Update cmd/mdatagen/validate_test.go

Co-authored-by: Mackenzie <63265430+mackjmr@users.noreply.github.com>

rogercoll committed:
Merge branch 'main' into add_podman_metadata

rogercoll committed:
Merge branch 'main' into add_podman_metadata

rogercoll committed:
Merge branch 'main' into add_podman_metadata

rogercoll committed:
Merge branch 'main' into add_podman_metadata

rogercoll commented:
@fatsheep9146 Is there anything missing for the merge?

fatsheep9146 commented:
> @fatsheep9146 Is there anything missing for the merge?

Sorry for delay, I got advise from maintainers @dmitryax about the unit, he suggested that we conform to the semantic conventions.

![image](https://github.com/open-telemetry/opentelemetry-collector-contrib/assets/11855957/644abcdb-d857-4cd3-86cb-e790c6963793)

I suggested we you `s` instead of `ns` first and open a pr in semantic convention related repo to try to prove the `ns` is better.
@rogercoll

rogercoll committed:
Merge branch 'main' into add_podman_metadata

rogercoll committed:
feat: move cpu usage metrics to seconds

Align with semantic conventions: https://opentelemetry.io/docs/specs/semconv/system/system-metrics/#metric-systemcputime

rogercoll commented:
@fatsheep9146 Cpu metrics precision changed in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/30232/commits/a5a00992ee43a646b6873608dbcfe31da59fafcb

fatsheep9146 commented:
@dmitryax could you help finally review about this pr?

dmitryax commented on .chloggen/add_podman_metadata.yaml:
```diff
@@ -0,0 +1,27 @@
+component: podmanreceiver
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: "Adds metrics and resources metadata and fixes invalid mapping of networking input and output metrics"
```
It's unclear from the description what metrics and metadata are added. Can we make the migration to metadata.yaml as a no-op change for the end user and handle any other changes separately?

rogercoll replied:
Make sense to me, let's make the "networking" changes into another PR. Removed in: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/30232/commits/91e24ff4bbbf8d199ab88b943c887781d27672c1

rogercoll committed:
Merge branch 'main' into add_podman_metadata

rogercoll committed:
rollback fix for networking tx/rx metrics

mx-psi commented on cmd/mdatagen/validate_test.go:
```diff
@@ -108,9 +108,17 @@ func TestValidate(t *testing.T) {
+		"container.cpu.utilization":           {"docker_stats", "kubeletstats"},
+		"container.cpu.usage.system":          {"docker_stats", "podman_stats"},
+		"container.cpu.usage.percpu":          {"docker_stats", "podman_stats"},
+		"container.cpu.usage.total":           {"docker_stats", "podman_stats"},
```
Is it okay to report the same metric with different units?

As discussed above, on the Docker stats receiver we have nanoseconds https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/facd3696decbc50cc906645cbd2a0d47f515c18c/receiver/dockerstatsreceiver/documentation.md?plain=1#L41-L47
while here we use seconds https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/91e24ff4bbbf8d199ab88b943c887781d27672c1/receiver/podmanreceiver/documentation.md?plain=1#L65-L71

Is that okay? If seconds is the right unit, shouldn't we use it on the Docker stats receiver as well?

rogercoll replied:
I don't think so, I would not use second's precision for the sake of convenience at the expense of precision. Instead, we should endeavor to establish distinct conventions specifically tailored to containers. Should we wait for the container's semantic convention https://github.com/open-telemetry/semantic-conventions/pull/282 (nanoseconds)?

mx-psi replied:
> I don't think so, I would not use second's precision for the sake of convenience at the expense of precision. Instead, we should endeavor to establish distinct conventions specifically tailored to containers.

My objection here is with the use of different units on each metric, I would expect them to have the same unit (whether it is nanoseconds or seconds I agree is something we can leave https://github.com/open-telemetry/semantic-conventions/pull/282 to decide on)

MovieStoreGuy commented:
Can I ask that the files in conflict get fixed up, outside of that, I don't see any outstanding issues that would block this PR?

fatsheep9146 commented:
> Can I ask that the files in conflict get fixed up, outside of that, I don't see any outstanding issues that would block this PR?

I think we still wait for the https://github.com/open-telemetry/semantic-conventions/pull/282 to be merged.

@MovieStoreGuy

fatsheep9146 commented:
just ping jsuereth on https://github.com/open-telemetry/semantic-conventions/pull/282 to merge the pr.

fatsheep9146 commented:
@rogercoll could you push forward this pr since https://github.com/open-telemetry/semantic-conventions/pull/282 is merge?

rogercoll committed:
Merge branch 'main' into add_podman_metadata

rogercoll committed:
Merge branch 'main' into add_podman_metadata

rogercoll committed:
Merge branch 'main' into add_podman_metadata

rogercoll commented:
After the merge of https://github.com/open-telemetry/semantic-conventions/pull/282, some additional changes will still be required:

- [ ] **IO**: From `container.blockio.io_service_bytes_recursive` to `container.disk.io` 
- [ ] **Network**: From `container.network.io.usage` to `container.network.io`
- [ ] **Memory**:
  - [ ] From `container.memory.usage.total` to `container.memory.usage`
  - [ ] Add `container.memory.limit` to semantic conventions
  - [ ] Add `container.memory.utilization` to semantic conventions
- [ ] **CPU**:
  - [ ] From `container.cpu.usage.system` and `container.cpu.usage.total` to `container.cpu.time`
  - [ ] Add `container.cpu.logical_number` attribute to semantic conventions.
  - [ ] Add `container.cpu.utilization` to semantic conventions.

MovieStoreGuy commented:
My preference would be to split the changes to keep change sizes small, are you okay with that @fatsheep9146 ?

fatsheep9146 commented:
> My preference would be to split the changes to keep change sizes small, are you okay with that @fatsheep9146 ?

I agreed. @rogercoll could you fix the conflicts, so I could push this pr to be merged ASAP.

rogercoll committed:
Merge branch 'main' into add_podman_metadata

rogercoll committed:
update changelog with breaking changes

rogercoll committed:
Merge branch 'main' into add_podman_metadata

rogercoll commented:
> I agreed. @rogercoll could you fix the conflicts, so I could push this pr to be merged ASAP.

Done, I also updated the changelog file to a "breaking" change type due to the cpu precision fix (ns -> s).

rogercoll commented:
If we prefer a totally no-op change PR, I could rollback the https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/30232/commits/a5a00992ee43a646b6873608dbcfe31da59fafcb commit and push it into a different PR. Let me know what you think
